src/main/java/org/victorrobotics/frc/dtlib/drivetrain/swerve/DTSwerveDrive.java

```diff
@@ -0,0 +1,154 @@
+        }
+        for (DTSwerveModule m : modules) {
+            Objects.requireNonNull(m);
+        }
+        this.modules = modules.clone();
+
+        setSubsystem("Swerve Drive");
+        setName(getSubsystem());
+
+        initializeHardware();
+
+        Translation2d[] wheelLocations = new Translation2d[modules.length];
+        positions = new SwerveModulePosition[modules.length];
+        for (int i = 0; i < wheelLocations.length; i++) {
+            wheelLocations[i] = modules[i].getLocation();
+            positions[i] = modules[i].getPosition();
+        }
+
+        kinematics = new SwerveDriveKinematics(wheelLocations);
+        poseEstimator = new SwerveDrivePoseEstimator(kinematics, getGyroAngle(), positions,
+                new Pose2d());
+
+        centerOfRotation = new Translation2d();
+        accelerationLimit = new DTAccelerationLimit();
+        currentSpeeds = new ChassisSpeeds();
+    }
+
+    public void initializeHardware() {
+        for (DTSwerveModule module : modules) {
+            module.initializeHardware();
+        }
+    }
+
+    public final void setCenterOfRotation(Translation2d newCenterOfRotation) {
+        centerOfRotation = newCenterOfRotation;
+    }
+
+    public void setAccelerationLimit(DTAccelerationLimit limit) {
+        accelerationLimit = limit;
+    }
+
+    public void setVelocityLimit(DTVelocityLimit limit) {
+        velocityLimit = limit;
+    }
+
+    public void setFieldRelative() {
+        isFieldRelative = true;
+    }
+
+    public void setRobotRelative() {
+        isFieldRelative = false;
+    }
+
+    protected abstract Rotation2d getGyroAngle();
+
+    public final void driveVelocity(double vx, double vy, double vr) {
+        driveVelocity(vx, vy, vr, centerOfRotation);
+    }
+
+    public void driveVelocity(double vx, double vy, double vr, Translation2d centerOfRotation) {
+        previousSpeeds = currentSpeeds;
+
+        if (isFieldRelative) {
+            currentSpeeds = ChassisSpeeds.fromFieldRelativeSpeeds(vx, vy, vr, getGyroAngle());
+        } else {
+            currentSpeeds = new ChassisSpeeds(vx, vy, vr);
+        }
+
+        velocityLimit.apply(currentSpeeds);
+        accelerationLimit.apply(currentSpeeds, previousSpeeds);
+
+        SwerveModuleState[] newStates = kinematics.toSwerveModuleStates(currentSpeeds,
+                centerOfRotation);
+        setStates(newStates);
+    }
+
+    public final void setStates(SwerveModuleState... states) {
+        if (states.length != modules.length) {
+            throw new DTIllegalArgumentException("received " + states.length + " module states for "
+                    + modules.length + " modules", (Object[]) states);
+        }
+
+        double minCosine = 1;
+        for (int i = 0; i < modules.length; i++) {
+            states[i] = SwerveModuleState.optimize(states[i], modules[i].getSteerAngle());
+            double cosine = Math.cos(Units.degreesToRadians(modules[i].getSteerAngle()
```

Comment:
Since we're dealing with degrees a lot for angle units, should we standardize on using `Rotation2d`? Then we can use `fromDegrees` or `fromRadians` as early as possible, so we're not passing around unitless doubles.

Reply:
I forgot that you can subtract two `Rotation2d`s and get their cosine, that’s a much better approach.

Reply:
As for standardizing on Rotation2d, yes, unless it would cause too much overhead from object creation / garbage collection.